[misc] minimize base image

das7pad opened the pull request:
With careful refactoring of image layers and additional configuration options for texlive, we can reduce the image size of the base image by about 322MB [1].

Once a file is included in an image layer, we can only hide it in the upper layers by deleting it. To not include an arbitrary file we need to delete it in the same layer it was created in - e.g. download the texlive installer, run the installer and delete it in the same RUN layer.

This PR drops these categories of artifacts: 
- `sudo` package - it was used to install node10, but as we run this as root we can omit sudo. 
- qpdf source
- texlive installer files, documentation and source files.
- apt package lists
- default nginx configuration - overwritten in the main stage
- npm download cache

In upgrading qpdf to a more modern version and pulling it as ubuntu package we can save another 38MB. The installation from source creates many artifacts we do not need, resulting in a 40MB layer. The ubuntu qpdf package is version 8.0.2 and has a footprint of 2MB on disk.
Given that we are only using the public api in clsi [2], this should not be an issue? This could be subject to a follow up PR.

---
[1] The image size is measured with `docker save IMAGE_ID | wc --bytes`. I am using the base image with tag `2.0-beta2` as reference. The image `sharelatex/sharelatex-base:2.0-beta-2` (`sha256:11ac166efa7067e8853fe6b6caa88637035a58a72d1e34c2130f178d85856cf4`) totals `1576994816` Bytes and an image build today using the proposed branch totals `1238641152` Bytes. That's a difference of 322MB.
[2] https://github.com/overleaf/clsi/blob/a62ff6e248d624c5ad78dbf08bb4613b043abdc2/app/coffee/OutputFileOptimiser.coffee#L35

das7pad committed:
[docker] drop sudo

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] install qpdf in a single stage

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] install texlive and additional tlmgr packages in a single stage

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] drop the apt package lists

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] pull the package lists only once

move the installation of nodejs into the dependencies install section

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] delete the default nginx configuration files immediately

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] skip the downloading and storage of unused texlive artifacts

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

das7pad committed:
[docker] drop the npm download cache

Signed-off-by: Jakob Ackermann <das7pad@outlook.com>

mserranom commented on Dockerfile-base:
```diff
@@ -9,56 +9,81 @@ ENV baseDir .
-RUN apt-get update
-RUN apt-get install -y sudo
-RUN apt-get install -y build-essential wget net-tools unzip time imagemagick optipng strace nginx git python zlib1g-dev libpcre3-dev aspell aspell-en aspell-af aspell-am aspell-ar aspell-ar-large aspell-bg aspell-bn aspell-br aspell-ca aspell-cs aspell-cy aspell-da aspell-de aspell-el aspell-eo aspell-es aspell-et aspell-eu-es aspell-fa aspell-fo aspell-fr aspell-ga aspell-gl-minimos aspell-gu as ...
+RUN apt-get update \
```
The main change in this `RUN` statement is cleaning up `/var/lib/apt/lists`, If that's the case I don't see the benefit to include node installation and nginx file deletion here.

das7pad replied:
The nodejs installation adds a package list and when splitting it into a second RUN stage, we would download the package lists twice.

nginx config files: see below.

mserranom replied:
ok, thanks!

mserranom commented on Dockerfile-base:
```diff
@@ -9,56 +9,81 @@ ENV baseDir .
+&&  apt-get install -y nodejs \
+    \
+&&  rm -rf \
+      /etc/nginx/nginx.conf \
```
A comment on the motivation to delete this files would be useful here. In the current `Dockerfile` the code is self explicative as new config files are added just afterwards.

das7pad replied:
They have to be deleted in the same layer in which they were created - otherwise they are still part of the final layer-stack. I can add a comment like the following. WDYT?
> We are adding a custom nginx config in the main Dockerfile.

mserranom replied:
looks good, thanks!

mserranom commented on Dockerfile-base:
```diff
@@ -9,56 +9,81 @@ ENV baseDir .
+      aspell aspell-en aspell-af aspell-am aspell-ar aspell-ar-large aspell-bg aspell-bn aspell-br aspell-ca aspell-cs aspell-cy aspell-da aspell-de aspell-el aspell-eo aspell-es aspell-et aspell-eu-es aspell-fa aspell-fo aspell-fr aspell-ga aspell-gl-minimos aspell-gu aspell-he aspell-hi aspell-hr aspell-hsb aspell-hu aspell-hy aspell-id aspell-is aspell-it aspell-kk aspell-kn aspell-ku aspell-l ...
+    \
+# install Node.JS 10
+&&  curl -sSL https://deb.nodesource.com/setup_10.x | bash - \
```
so `sudo` is not needed here? Interesting

mserranom commented on Dockerfile-base:
```diff
@@ -9,56 +9,81 @@ ENV baseDir .
+&&    make \
+&&    make install \
+&&    ldconfig \
+    \
```
Just personal taste, so ignore as you wish: I find this backslash-only lines a bit noisy, indentation works just well for me with `&& \` at the end of the line.

Is this a particular type of indenting/styling Docker files? It's pretty much a redo of our formatting

das7pad replied:
The idea here was to split the different tasks into separate blocks: environment setup, downloading, compiling and cleanup.

When using the debian package, these steps can be dropped.

mserranom reviewed (changes requested):
added several comments

mserranom commented on Dockerfile-base:
```diff
@@ -9,56 +9,81 @@ ENV baseDir .
-ENV PATH /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/2019/bin/x86_64-linux/
-RUN tlmgr install latexmk
-RUN tlmgr install texcount
+ARG TEXLIVE_MIRROR=http://mirror.ctan.org/systems/texlive/tlnet
```
Can the comment on `CTAN mirrors occasionally fail,...` be added back?

das7pad replied:
Sure. I will add an example shell command with `--build-arg` flags too.

mserranom commented:
Upgrading to `qpdf` `8.0.2` should be fine, we're using it on Texlive images https://github.com/overleaf/texlive-images/blob/e1de3fa0dd2ed156440521b44641f9a6dac9e25a/texlive-full/2017.1/Dockerfile.2017.upgrade-qpdf#L1

das7pad committed:
[docker] apply review feedback

- install qpdf as ubuntu package
- add a comment on the nginx config removal
- add back and update a note on changing the texlive mirror

mserranom reviewed (approved):
Looks good, thanks!

See comments on https://github.com/overleaf/docker-image/pull/119#issuecomment-579192912: we're putting this PR in held until the next release is out. I'll keep this assigned to me until merging.

mserranom commented:
tested, all looks good.
